[NETBEANS-377] Add PSR-4 hints

KacerCZ opened the pull request:
https://issues.apache.org/jira/browse/NETBEANS-377

- Added hints for [PSR-4 Autoloader](https://www.php-fig.org/psr/psr-4/) standard.
- Checks that class/interface/trait declaration matches file name.
- Checks that namespace matches directory structure.
- Highlights only non-matching part of namespace.

junichi11 commented:
We should leave a comment for unsupported cases somewhere.

Add test for the following case?
``` php
// file: Vendor/App/ClassName.php
namespace Vendor\App;

class classname {

}

```

junichi11 commented on php/php.editor/src/org/netbeans/modules/php/editor/verification/PSR4Hint.java:
```diff
@@ -0,0 +1,258 @@
+        }
+
+        public List<Hint> getHints() {
+            return hints;
```
Use `Collections.unmodifiableList(hints)`?

KacerCZ commented:
I added tests where casing is different.

Also added note about case-sensitivity to hint descriptions:
- "Namespace MUST match (case-sensitive) file directory." 
- "Type name MUST match (case-sensitive) file name."

tmysik commented on php/php.editor/src/org/netbeans/modules/php/editor/verification/Bundle.properties:
```diff
@@ -31,6 +31,7 @@ ImmutableVariablesCustomizer.jLabel1.text=Number of Allowed Assignments per Vari
 csl-hints/text/x-php5/hints/braces=Braces
 csl-hints/text/x-php5/hints/psr0=PSR-0: Autoloading Standard
 csl-hints/text/x-php5/hints/psr1=PSR-1: Basic Coding Standard
+csl-hints/text/x-php5/hints/psr4=PSR-4: Autoloader
```
Should not it be "Autoloading Standard", similarly to PSR-0?

KacerCZ replied:
"Autoloader" is official name of PSR-4 - https://www.php-fig.org/psr/psr-4/

tmysik replied:
OK.

tmysik commented on php/php.editor/src/org/netbeans/modules/php/editor/verification/PSR4Hint.java:
```diff
@@ -0,0 +1,259 @@
+                    int endOffset = namespaceName.getEndOffset();
+                    File currentDir = getFile().getParentFile();
+                    ListIterator<Identifier> segmentsIterator = namespaceName.getSegments().listIterator(namespaceName.getSegments().size());
+                    while (segmentsIterator.hasPrevious()) {
```
Perhaps calling `CancelSupport.getDefault().isCancelled()` in this loop could be useful?

KacerCZ replied:
One iteration of the loop will be done for one part of namespace.
I expect no more than 10 iterations even if someone uses very deep structure.

Can traversing directories can take long time and getting its name take so much time it will be noticeable?

tmysik replied:
In general, this check should be present in every loop to avoid unneeded work to be done. The check is very cheap and IO operations are slow, in general (yes, I heard about SSD :) so I think it should be there.

KacerCZ replied:
Right, I added it.

tmysik replied:
Thanks!

junichi11 commented:
Will re-run jobs.

KacerCZ commented:
@junichi11 @tmysik Thanks for review.

ahajzik-crido commented:
Does this feature support psr-4 declaration from composer.json?

KacerCZ commented:
@ahajzik-crido No, mapping from composer.json is separate task - https://issues.apache.org/jira/browse/NETBEANS-5341
